[Demo] added an end-user oriented demonstration of itowns capabilities

Nynjin opened the pull request:
## Description
This pull request introduces a new demonstration example for iTowns, adding a complete demo application under `examples/demo`, integrated amongst the other Itowns examples under the section `Demo`.

This demo section features multiple self-contained scenes with navigation, camera transitions, and layer management (visibility, controls, feature picking, recreation of view and layers from scratch, ...).

The `examples/demo` package can be described as such :
> `css/` : style for the UI elements
>
> `src/` : Typescript demo code
>
>> `Config/` : global constants such as scene transition duration or distance from globe where to toggle on/off atmosphere
>>
>> `Layers/` : objects containing promises for itowns layered shared between multiple scenes to avoid refetching/re-instantiation, as well as methods such as feature picking which is specific to each layer
>>
>> `Sources/` : promises containing sources that are used by multiple layers to avoid refetching
>>
>> `Views/` : singletons for the different views shared between layers (globe, immersive, planar, ...) that initialize itowns views and expose extra methods/properties (visibility toggle, viewerDiv, layer insertion, instance clearing, ...)
>>
>> `Scenes/` : scenes used in the demo containing their methods and properties (title, description, layers, onCreate, onEnter, onExit, view)
>>
>> `Repositories/` : shared singleton containers such as SceneRepository (array of all used scenes)
>>
>> `Types/` : common types and wrappers (SceneType, or LayerType containing all types of itowns layers used)
>>
>> `Utils/` : global methods such as scene transition, ui update, camera movement, event propagation block, ...
>>
>> `Services/` : application-level services such as feature picking on click
>>
>> `index.ts` : main demo entrypoint
>
> `index.html` : entrypoint for the demo. Hosts the root containers for the UI


## Motivation and Context
iTowns currently provides an unstructured collection of examples that are mainly designed for developers and are kind of tutorials in the end. This PR adds an end-user oriented demonstration of main iTowns features.
See [related issue](https://github.com/iTowns/itowns.github.io/issues/35).

### Environment

- OS: Windows 11
- Browser: Brave
- Node: v24.11.0
- npm: 11.6.1

## Screenshots

https://github.com/user-attachments/assets/d7fa247e-3531-4450-a7db-583abb61ca0e

| Globe Scene | Planar Scene | Terrain Elevation Scene |
| --- | --- | --- |
| <img src="https://github.com/user-attachments/assets/680908fc-596a-4e57-8835-007554c26c69" width="350" /> | <img src="https://github.com/user-attachments/assets/ce7d85dc-8a6a-44d0-b7f8-8397dc627839" width="350" /> | <img src="https://github.com/user-attachments/assets/84be7d6b-5275-4d79-8edd-95b7b794e58f" width="350" /> |

| Instanced Data Scene | Point Cloud Scene | BIM Scene |
| --- | --- | --- |
| <img src="https://github.com/user-attachments/assets/25f6fad0-ad64-444e-a299-2fd30408b675" width="350" /> | <img src="https://github.com/user-attachments/assets/5fc332f3-6f17-402e-b1d0-fc5578aa12c6" width="350" /> | <img src="https://github.com/user-attachments/assets/ebbbf08d-c306-4d9c-8f76-ebe267ffdcfa" width="350" /> |

| Textured Meshes Scene | Immersive Scene |
| --- | --- |
| <img src="https://github.com/user-attachments/assets/1dd3d70d-67ce-41c8-95ae-ace55e6efaa0" width="350" /> | <img src="https://github.com/user-attachments/assets/01771bdf-bcdb-49a5-9d25-1eb88c1dcf7f" width="350" /> |

## Open questions / TODO
* Find a better 3DTiles dataset
* Find a better point cloud (bigger, with classification and RGB if possible). Which dataset can we use from the LIDAR HD ? 
* Immersive and planar scene tend to be unstable (white screen sometimes, ground on immersive scene may not appear), mostly fixed with scene reset
* Add extruded lines when ready
* Move light on globe scene to not obscure the center
* Camera movement based on duration can be troublesome when switching between scenes that have the same placement, or if resetting camera position when it hasn't moved, or has barely moved (waiting 2 seconds each time and blocking the UI)

Nynjin committed:
feat(demo): add demo section with multiple scenes and transitions

camera movement between scene coordinates, layers removal/additions, onEnter/onExit for extra properties

Nynjin committed:
feat(demo): sceneManager with navigation + added tilt to scenes

Nynjin committed:
feat(demo): added transition between 3D View and Immersive View + manage shared layers between multiple scenes

Nynjin committed:
fix(demo): render the same scene twice + scene heading

- instead of removing / readding same layers which is bug-prone, don't remove and just toggle visibility

- added heading to scenes to ensure camera ends up in the right spot

Nynjin committed:
chore(demo): typescript migration

Nynjin committed:
feat(demo): added scene ready state + allow scene to have different coordinate projection

- onCreate promise for scene first instantiation to set ready state
- onEnter for each entry in scene
- Scenes manage their layers to handle dependencies, transitionToScene only manages visibility
- better ImmersiveView types

Nynjin committed:
fix(demo): prevent camera controls during scene transition

Nynjin committed:
feat(demo): updated 2D projected data and 3D extruded data scenes to fit specification with same pov and shared layers

Nynjin committed:
feat(demo): added InstancedData scene with trees in Lyon using GLTF loader

Nynjin committed:
feat(demo): PointCloud scene representing Part Dieu station

added PointCloudView with PlanarControls, fixed default camera position for ImmersiveView + added IgnMnt elevation layer where missing

Nynjin committed:
feat(demo): added intermediate view for cleaner camera movements between scenes

Nynjin committed:
feat(demo): added scene title and description + reset scene and reset demo

Nynjin committed:
fix(demo): UI updates and navigation buttons

fixed empty UI on first scene, UI updates before camera transition now + navigation buttons now have fixed position regardless of text

Nynjin committed:
fix(demo): smoother camera transitions

Transition now uses camera's coordinates instead of current scene coordinates + avoid unecessary unzoom step when next scene's position is already further away that computed minimal distance for transition

Nynjin committed:
fix(demo): Point Cloud scene

- fixed config duplicates on enter
- fixed insertion of PointCloud layer in view to not use a parentLayer
- removed unecessary layers
- Point Cloud now appears on 3D Globe

Nynjin committed:
fix(demo): pointCloudLayer remaining visible after transition

gui needs to be hidden before layer visibility is set to false

Nynjin committed:
fix(demo): duplicate layers adding to view

- when navigating too fast between multiple scenes that have shared layers, async adding same layer to view could happen in parallel and cause error

Nynjin committed:
feat(demo): added tooltips for park layer and config GUI to scene

one GUI per view with transition between views and management of Scenes with a specific GUI (e.g. PointCloud)

Nynjin committed:
feat(demo): add 3D textured meshes scene

Nynjin committed:
feat(demo): added planar view of lyon

PlanarView requires extent in constructor
so instead of singleton like other views, there is one instance per extent

Nynjin committed:
fix(demo): navigation buttons fixed position

Nynjin committed:
fix(demo): Layer typing

added custom type to encapsulate all currently required itowns layer types to avoid ts error

Nynjin committed:
perf(demo): scene transition layer management

updated scene transition to avoid displaying current scene's layers and next scene's layers at the same time
minor fix on gui transition to remove elevation and geoid layers as well

Nynjin committed:
feat(demo): BIM visualization scene

Nynjin committed:
feat(demo): compatibility with three-geospatial lighting and atmosphere

- removed unecessary ambiant lights
- added atmosphere removal for scenes that do not require it with transition based on scene ranges

Nynjin committed:
revert(demo): remove debug UI and bugged featureInfo

Nynjin committed:
feat(demo): improve soft reset and added hard reset for scene

- refactor required to make layer promises and cache accessible to hard reset method by turning layers to objects with layerRepository
- added clearInstance method to View in order to remove viewerDiv, dispose view and delete instance
- moved sceneRepository outside of index.html to make it accessible
- refactor on some types and scene names to make it clearer

Nynjin committed:
feat(demo): atmosphere added/removed automatically on distance from globe instead of travel time and boolean

Nynjin committed:
feat(demo): replaced forward/backward navigation with one button per scene

Nynjin committed:
fix(demo): pointer events blocked by demo ui

Nynjin committed:
fix(demo): scene reset

- prevent reset while doing scene transition
- bypass view dispose error if possible

Nynjin committed:
feat(demo): better text contrast

Nynjin committed:
fix(demo): layer dissapearing on scene transition

- use Globe3dScene as transition between scenes with its layers visible
- prevent actions during scene transition

Nynjin committed:
feat(demo): update scene titles and descriptions + scene order

Nynjin committed:
fix(demo): elevation layer artefacts

- replaced IGN_MNT elevation layer by WORLD_DTM

Nynjin committed:
fix(demo): error after ImmersiveView reset scene

- prevent moveCameraTo if not a globe view
- reset transition scene as well

Nynjin committed:
feat(demo): moved entire demo UI to center-bottom + extra zoom on TexturedMeshes scene

Nynjin committed:
feat(demo): added Feature Picker

- display informations about an element from a tracked layer
- each LayerPromise manages how their info are extracted/structured with getPickingInfo

Nynjin committed:
fix(demo): View3d camera teleporting to scene range

Nynjin committed:
feat(demo): added Point Cloud controls

- switch between color and classification render with buttons

Nynjin committed:
fix(demo): feature picker interaction remaining when layers invisible

- filter registered layers by visibility before feature picking

Nynjin committed:
fix(demo): prevent events to propagate from UI elements to itowns view

- made FeaturePickerService's blockEvents method a global utils so PointCloudScene's UI can use it
- check skymanager's existence and enabled property before define

Nynjin committed:
fix(demo): GlobeView skymanager property not found by typescript

- use custom type to add skymanager to view

Nynjin committed:
fix(demo): remove featurePicking UI when in a scene that doesn't use it

Nynjin committed:
chore(demo): update demo deps, exports, and webpack config

- Added "itowns" as a dependency in the demo's package.json
- Refactored modules with index.ts exports for each namespace
- Removed LayerRepository and replaced its usage with direct layer imports
- Updated the webpack configuration to support the demo example

Desplandis commented:
I would not be able to do a review until new year's eve (vacation time :D), @alavenant @Calvina24 could you review the functionality of this PR and dispatch the technical part to a member of the team?

Desplandis commented:
@Nynjin I saw that there is a 10mB binary file in the changes, would it be possible to remove it from the commits and host it elsewhere to not pollute the main itowns repository? We usually use [itowns2-sample-data](https://github.com/iTowns/iTowns2-sample-data) for the datasets.

Nynjin commented:
> @Nynjin I saw that there is a 10mB binary file in the changes, would it be possible to remove it from the commits and host it elsewhere to not pollute the main itowns repository? We usually use [itowns2-sample-data](https://github.com/iTowns/iTowns2-sample-data) for the datasets.

Got it thanks, I uploaded my models there and made a [PR](https://github.com/iTowns/iTowns2-sample-data/pull/21).
Once that's done I'll update this one to remove the local models and use the uploaded ones instead.

alavenant commented:
Thanks for the work. 

The transition from one demo to another is smooth and enjoyable.

I suggest to put the "go local" next to "bring your data to life". 

In "Feel the Relief", you could also talk about the ortho-images.

"go local" : unzoom so that we can fully appreciate the scene as a whole

Why not add a view "combine your data" with lidar, 3D objects, buildings, bim, ... ?

Nynjin commented:
> Thanks for the work.
> 
> The transition from one demo to another is smooth and enjoyable.
> 
> I suggest to put the "go local" next to "bring your data to life".
> 
> In "Feel the Relief", you could also talk about the ortho-images.
> 
> "go local" : unzoom so that we can fully appreciate the scene as a whole
> 
> Why not add a view "combine your data" with lidar, 3D objects, buildings, bim, ... ?

Thank you for you feedback.

I did consider putting "go local" before "bring your data to life" but having a planar view right in the middle of globe views looks a bit off, that's mostly why planar and street views are at the very end.

And when it comes to combining the data, it's an interesting idea but I'd have to check how it holds up performance-wise since clicking directly on that scene would attempt to instantiate all of these elements at the same time.

mgermerie commented on examples/demo/index.html:
```diff
@@ -0,0 +1,162 @@
+<!DOCTYPE html>
+<html>
+    <head>
+        <title>Itowns - Demonstration</title>
```
A minor detail but maybe "Showcase" would be more suited than "Demonstration" ?

Nynjin replied:
Showcase does sound better, I renamed it

mgermerie commented on examples/demo/src/Scenes/Globe3dScene.ts:
```diff
@@ -0,0 +1,33 @@
+        heading: 0,
+    },
+    layers: [],
+    view: new View3D(),
```
This should not be necessarily initialized with an instance of `View3D` since it is later overwritten within `onCreate` method (on line 23).

mgermerie replied:
And the same goes for all other scenes.

Nynjin replied:
You're right, I changed it to undefined until onCreate initializes it and I added getters to retrieve the view

mgermerie commented on examples/demo/src/Views/View.ts:
```diff
@@ -0,0 +1,77 @@
+        }
+
+        const layerPromises = layers.map(async (layer) => {
+            if (this.view!.getLayerById(layer.id)) {
```
The non-null assertion here should not be necessary since `this.view` was already checked on line 58.

mgermerie replied:
The same goes for line 67.

Nynjin replied:
It happens because the non-null assertions was outside of the promise. And by the time the promise is executed, typescript considers it could be null again. I just moved the assertion inside the promise

mgermerie commented on examples/demo/src/Scenes/InstancedDataScene.ts:
```diff
@@ -0,0 +1,79 @@
+        }
+        InstancedDataScene.view = new View3D();
+
+        const view = InstancedDataScene.view.getView();
```
For clarity, perhaps you could rename this constant to `itownsView`, for example. This would help distinguish between the `View3D` and `itowns.GlobeView` types.

Nynjin replied:
You're right it's clearer. I renamed the "getView" method to "getItownsView" as well

mgermerie commented on examples/demo/src/Sources/index.ts:
```diff
@@ -0,0 +1 @@
+export * as BuildingsSource from './BuildingsSource';
```
I understand that this data source is the only one used in multiple layers, which is probably why it’s the only one exposed in `Sources/`.

However, for better code consistency, I suggest declaring all sources in `Sources/`, even if they are used by only a single layer. In my opinion, this would serve two purposes:
- If a source needs updating, it’s more straightforward to find the line to change in the `Sources/` folder rather than inside a specific file in `Layers/`.
- The showcase might evolve and eventually display the same 3D tiles set in multiple layers across different scenes, for instance.

Then there is the issue of ortho imagery and DEM sources, whose parameters are obtained from fetching JSON files. This approach is used in the current examples, but I don’t think it should be replicated elsewhere - it only serves to make the example code shorter by removing lines that aren’t relevant to the feature being demonstrated.
In my opinion, it would be better to copy these source configurations into specific files in the `Sources/` folder, keeping the `demo` and `examples` packages self-contained.

Nynjin replied:
Got it, I separated source loading, 3D model loading and layers
In my opinion the JSON fetcher can still be interesting for code examples on using config files so for now I just copied the json configuration to be inside the demo in order to keep it self contained
But it could be removed to use the configuration directly in the source

mgermerie commented on examples/demo/src/Utils/SceneTransitionUtils.ts:
```diff
@@ -0,0 +1,250 @@
+            }
+        }
+
+        FeaturePickerService.container!.style.display =
```
On some of the transitions I tested, `FeaturePickerService.container!.style` was `null`, causing an error when reading `display`.

Nynjin replied:
I added an extra check to make sure FeaturePickerService container is defined before using it

mgermerie commented on examples/demo/src/Views/View.ts:
```diff
@@ -0,0 +1,77 @@
+        }
+
+        this.viewerDiv.setAttribute('id', visible ? 'viewerDiv' : this.id);
+        this.viewerDiv.style.display = visible ? 'block' : 'none';
```
From my understanding, simply setting `display: none` on a three.js canvas container does not prevent animation frames from being requested. It therefore continues to use resources unnecessarily.

You could try:
```js
this.view.tileLayer.frozen = !visible;
this.view.notifyChange();
```
but I’m not sure whether there is a better way to prevent three.js from rendering. @Desplandis or @HoloTheDrunk, do you have any insight on this ?

Desplandis replied:
Since we lazily request animation frames, if we do not trigger update (no call to `view#notifyChange` & co) we should not update / render another frame!

mgermerie commented on webpack.config.cjs:
```diff
@@ -158,8 +159,36 @@ module.exports = () => {
+        },
+        resolve: {
+            ...configESM.resolve,
+        },
```
As `output` and `resolve` are already properties of `configESM`, you should not need to overwrite them here. They should already be accounted for with the spread on line 163.

Nynjin replied:
You're right, I removed the redundant overwrite

mgermerie reviewed (commented):
Thank you for this PR! I found it well documented, with clearly written and well-architected code, which made it pleasant to review.

I left some comments below. If I understand correctly, the UI of the demo will change. Will this be done in this PR or in another one?

Nynjin committed:
refactor(demo): scene management and layer handling

- Updated ThreeJS dependency in demo
- Renamed demo to Showcase
- Avoid reinstantiation of scene view on create
- Rename iTowns view getter to be more distinct from Scene view
- Separated Scenes, layer creation and source fetching for all layers
- Updated demo Webpack configuration to remove redundancies
- Atmophere handling fix after rebase on latest iTowns version

Nynjin committed:
feat(demo): added combined data scene + gltf loading refactor

- combined data scene with parks, buildings, point cloud and BIM
- separated model loaders from layers and scenes
- prevent camera animation if scenes have the same coordinates
- copied used json layers to demo assets

Nynjin committed:
refactor(demo): specify if object is a demo view or an itowns view

Nynjin commented:
> > Thanks for the work.
> > The transition from one demo to another is smooth and enjoyable.
> > I suggest to put the "go local" next to "bring your data to life".
> > In "Feel the Relief", you could also talk about the ortho-images.
> > "go local" : unzoom so that we can fully appreciate the scene as a whole
> > Why not add a view "combine your data" with lidar, 3D objects, buildings, bim, ... ?
> 
> Thank you for you feedback.
> 
> I did consider putting "go local" before "bring your data to life" but having a planar view right in the middle of globe views looks a bit off, that's mostly why planar and street views are at the very end.
> 
> And when it comes to combining the data, it's an interesting idea but I'd have to check how it holds up performance-wise since clicking directly on that scene would attempt to instantiate all of these elements at the same time.

@alavenant I added a `Combine Your Data` scene with the data from parks, point cloud, buildings, 3D trees and BIM
It's pretty slow when using performance and network throttling but if you navigate the scenes one by one then all of the layers are already cached and it's nearly instant
I just used the data I already had to see how it'd look so the BIM model is out of place. If there's a proper dataset it could be interesting

<img width="480" height="270" alt="image" src="https://github.com/user-attachments/assets/2b8ac140-5610-48e4-b940-4676e8beebd6" />

mgermerie reviewed (commented):
Thanks for the updates.

I just tested the demo again and found a small glitch: if you change the point cloud color option and then switch scenes, the UI for that option is not updated accordingly. For example, you can toggle the option to "Color" in `PointCloudScene` and then switch to `CombinedDataScene`. The UI will reset to "Classification", while the points themselves remain in "Color" mode.

Other than that, it looks all good to me.

Would you mind squashing the commits as well ? Since this PR introduce a single large feature (the showcase addition), it might be clearer from a changelog/history perspective to keep it as one commit. That said, it would make for quite a large commit, so I'd be interested in hearing your thoughs on this as well, @Desplandis and @jailln.

jailln commented:
> Thanks for the updates.
> 
> I just tested the demo again and found a small glitch: if you change the point cloud color option and then switch scenes, the UI for that option is not updated accordingly. For example, you can toggle the option to "Color" in `PointCloudScene` and then switch to `CombinedDataScene`. The UI will reset to "Classification", while the points themselves remain in "Color" mode.
> 
> Other than that, it looks all good to me.
> 
> Would you mind squashing the commits as well ? Since this PR introduce a single large feature (the showcase addition), it might be clearer from a changelog/history perspective to keep it as one commit. That said, it would make for quite a large commit, so I'd be interested in hearing your thoughs on this as well, @Desplandis and @jailln.

One commit is fine by me :+1:

Nynjin committed:
fix(demo): issues with new examples page interface & point cloud UI sync

- relative paths fixed
- demo webpack config eslint defaulting to flat and causing error
- css style breaking
- point cloud mode switch (color / classification) fix so the active button stays synced with current point cloud rendered mode

Nynjin commented:
> Thanks for the updates.
> 
> I just tested the demo again and found a small glitch: if you change the point cloud color option and then switch scenes, the UI for that option is not updated accordingly. For example, you can toggle the option to "Color" in `PointCloudScene` and then switch to `CombinedDataScene`. The UI will reset to "Classification", while the points themselves remain in "Color" mode.
> 
> Other than that, it looks all good to me.
> 
> Would you mind squashing the commits as well ? Since this PR introduce a single large feature (the showcase addition), it might be clearer from a changelog/history perspective to keep it as one commit. That said, it would make for quite a large commit, so I'd be interested in hearing your thoughs on this as well, @Desplandis and @jailln.

Thank you for the review.

I fixed the UI to stay synced with the point cloud layer + merge conflicts as I just rebased.
I can squash the commits during the merge.

Nynjin committed:
feat(demo): updated point cloud dataset and camera angle

AnthonyGlt commented:
@tim-btrn

Nynjin committed:
fix(demo): three-geospatial atmosphere enabling/disabling

Nynjin commented:
Regarding the Point Cloud view, I've switched to the dataset used in `entwine-simple-loader` which isn't lidar-hd and has stray points in the sky, though it seems to look a little better as it is much larger and has colors. 
It would probably be a lot better with EDL but for now it looks like this :
<img width="640" height="344" alt="image" src="https://github.com/user-attachments/assets/1098e207-9446-4bfd-97ad-e366e9e3323a" />
<img width="640" height="344" alt="image" src="https://github.com/user-attachments/assets/c60752d3-9e70-4cf4-a320-ce0cabe2eca7" />

Nynjin committed:
fix(demo): asset paths when accessing html directly + black screen on changing view type

- when instancing new view, potential conflicts if current view is different
- asset paths set in config when html renders to work on both examples page and direct access
- fix point cloud being defaulted back to color on first render in different scenes

mgermerie reviewed (approved):
Thanks for this much-needed addition! I understand that the UI changes you discussed with your designers will come later in another PR. Is that right ?

Nynjin commented:
> Thanks for this much-needed addition! I understand that the UI changes you discussed with your designers will come later in another PR. Is that right ?

That's right. @tim-btrn will be the one working on it.
